mysql/datadog_checks/mysql/queries.py

```diff
@@ -74,6 +74,87 @@
 SELECT plugin_status
 FROM information_schema.plugins WHERE plugin_name='group_replication'"""
 
+SQL_DATABASES = """
+SELECT schema_name as `name`,default_character_set_name,default_collation_name FROM information_schema.SCHEMATA
+WHERE schema_name not in ('mysql', 'performance_schema', 'information_schema')"""
+
+SQL_TABLES = """\
+SELECT table_name as `name` FROM information_schema.TABLES WHERE TABLE_SCHEMA = %s
+"""
+
+SQL_COLUMNS = """\
+SELECT table_name,
+       column_name as `name`,
+       data_type, column_default as `default`,
+       is_nullable as `nullable`,
+       ordinal_position
+FROM INFORMATION_SCHEMA.COLUMNS
+WHERE table_schema = %s AND table_name IN ({});
+"""
+
+SQL_INDEXES = """\
+SELECT
+    table_name,
+    index_name as `name`,
+    collation,
+    cardinality,
+    index_type,
+    group_concat(seq_in_index order by seq_in_index asc) as seq_in_index,
+    group_concat(column_name order by seq_in_index asc) as columns,
+    group_concat(sub_part order by seq_in_index asc) as sub_parts,
+    group_concat(packed order by seq_in_index asc) as packed,
+    group_concat(nullable order by seq_in_index asc) as nullables,
+    group_concat(non_unique order by seq_in_index asc) as non_uniques
+FROM INFORMATION_SCHEMA.STATISTICS
+WHERE table_schema = %s AND table_name IN ({})
+GROUP BY table_name, index_name, collation, cardinality, index_type;
+"""
+
+SQL_FOREIGN_KEYS = """\
+SELECT
+    constraint_schema,
+    constraint_name as `name`,
+    table_name,
+    group_concat(column_name) as column_names,
+    referenced_table_schema,
+    referenced_table_name,
+    group_concat(referenced_column_name) as referenced_column_names
+FROM
+    INFORMATION_SCHEMA.KEY_COLUMN_USAGE
+WHERE
+    table_schema = %s AND table_name in ({})
+    AND referenced_table_name is not null
+GROUP BY constraint_schema, constraint_name, table_name, referenced_table_schema, referenced_table_name;
+"""
+
+SQL_PARTITION = """\
+SELECT
+    table_name,
+    partition_name as `name`,
+    group_concat(subpartition_name order by subpartition_ordinal_position asc) as subpartition_names,
```

Comment:
re: https://github.com/DataDog/integrations-core/pull/17916/files#r1662586706

Overall, the concatenation approach seems like a hack. We'll have to coordinate the un-concatenating logic across the backend and agent, which makes changes difficult in the long run. The tradeoff of sending less data is a micro-optimization if the payloads are compressed. 

e.g. here is an experiment (via chatgpt) showing that across gzip compressed payloads, there is a negligible difference between concatenating nested objects versus nesting the full JSON for sub-partitions

```python
import json
import gzip
import random
import string
import sys

# Function to generate JSON objects with consistent schema
def generate_json(num_entries):
    data = []
    for _ in range(num_entries):
        entry = {
            "name": ''.join(random.choices(string.ascii_letters, k=10)),
            "age": random.randint(20, 70),
            "city": ''.join(random.choices(string.ascii_letters, k=10))
        }
        data.append(entry)
    return data

# Function to transform JSON by removing field names
def transform_json(data):
    transformed = [[entry["name"], entry["age"], entry["city"]] for entry in data]
    return transformed

# Function to calculate size of JSON data in bytes
def calculate_size(data):
    return sys.getsizeof(json.dumps(data))

# Function to calculate gzip compressed size
def calculate_gzip_size(data):
    json_data = json.dumps(data).encode('utf-8')
    return len(gzip.compress(json_data))

# Generate data of different sizes
sizes = [10, 100, 1000, 10000]
results = []

for size in sizes:
    original_data = generate_json(size)
    transformed_data = transform_json(original_data)
    
    original_size = calculate_size(original_data)
    transformed_size = calculate_size(transformed_data)
    
    original_gzip_size = calculate_gzip_size(original_data)
    transformed_gzip_size = calculate_gzip_size(transformed_data)
    
    results.append({
        "num_entries": size,
        "original_size": original_size,
        "transformed_size": transformed_size,
        "original_gzip_size": original_gzip_size,
        "transformed_gzip_size": transformed_gzip_size
    })

results
```


Number of Entries | Original Size (bytes) | Transformed Size (bytes) | Original Gzip Size (bytes) | Transformed Gzip Size (bytes)
-- | -- | -- | -- | --
10 | 619 | 389 | 291 | 265
100 | 5749 | 3449 | 2058 | 1986
1000 | 57049 | 34049 | 19090 | 18613
10000 | 570049 | 340049 | 188373 | 184022

Reply:
There is also a potential bug here in using `group_concat()`. Because this function only concatenates _non-null_ values and your solution depends on the ordering of concatenated values, a null value in the rows will break the ordering. I don't know if that will happen in practice, but it's hard to test for and a risk we should try to avoid.

Reply:
"The tradeoff of sending less data is a micro-optimization if the payloads are compressed." -  thank you for sharing this, as you saw on our one on one I totally didn't think about it.

"We'll have to coordinate the un-concatenating logic across the backend and agent, which makes changes difficult in the long run." - I didn't plan to un-concatenate. I thought we might even want to compress further. 

What I mean is that so far our approach was to simplify and concatenate data on the REDAPL side. We don't have as many tables as Postgres or SQLServer to describe columns , indexes , partitons and etc.. When it comes to partitions we have only 
[https://github.com/DataDog/infrastructure-resources/blob/5b94019a4ecec4d8d6ef32f3956d63c201052414/generated/yaml/postgres_table.yaml#L44] and the number of partitions. 

Like for indexes if index is defined on several columns we would have 
[https://github.com/DataDog/infrastructure-resources/blob/5b94019a4ecec4d8d6ef32f3956d63c201052414/generated/yaml/sqlserver_table.yaml#L54](url)
a concatenation of columns.

I just wanted to share it for a sync, regardless I can send a field for each sub-partition and then concatenate them on the backend.

Reply:
Hmm I'm not sure how I feel about the concatenation. Especially with the behavior I described with `group_concat`:  https://github.com/DataDog/integrations-core/pull/17916#discussion_r1666174717. 

Tbh there's a risk here of overcomplicating the implementation for a feature which I suspect is not even widely used (sub-partitioning). The more I think about it, the more I lean towards removing sub-partitioning for now. I.e. we will only store partitions and add an attribute like `has_suppartitions`. Then if we see customers using subpartitions with frequency, we can add support for them explicitly. What do you think of this approach?

Reply:
May be take a look at the current implementation - I removed `group_concat` this morning. We can always simplify on the backend to `subpartition_count`.

Reply:
> which I suspect is not even widely used (sub-partitioning). 
I saw sub partitioning being used in practice a lot in large Oracle DWH databases. For example, partition by date, and then list or hash subpartitions on some other fields. Not sure how relevant is this observation for MySQL. In interest of agility, I'm supporting the idea of making the first version without subpartitions, and add them later if requested by customers.

Reply:
Looks good with the updates. If needed, we can collapse subpartitions on the backend.